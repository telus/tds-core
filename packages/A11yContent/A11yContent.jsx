import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { safeRest } from '@tds/util-helpers'
import { componentWithName, or } from '@tds/util-prop-types'

const StyledA11yContent = styled.span({
  position: 'absolute',
  height: '1px',
  width: '1px',
  overflow: 'hidden',
  clip: 'rect(1px, 1px, 1px, 1px)',
})

/**
 * @version ./package.json
 */

const A11yContent = ({ children, ...rest }) => {
  return <StyledA11yContent {...safeRest(rest)}>{children}</StyledA11yContent>
}

A11yContent.propTypes = {
  /**
   * Accessible content. Can be either a string or a `Heading` Component
   */
  children: or([PropTypes.string, componentWithName('Heading')]).isRequired,
}
A11yContent.defaultProps = {}

export default A11yContent
