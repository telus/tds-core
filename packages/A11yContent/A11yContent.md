Allows content to be hidden from sight, but still visible to screen readers. This can be used to give context to users using screen readers without revealing unnecessary extra text to sighted users.

```jsx
    <Text>The screen reader says:</Text><A11yContent>This is accessible content</A11yContent>
```

If `A11yContent` has to go between two separate text components, considering wrapping the whole text in a `<span>` to persist leading and trailing space.

```jsx
<ButtonLink href="#">
  <span>
    Shop <A11yContent>iPhone</A11yContent> Now
  </span>
</ButtonLink>
```
