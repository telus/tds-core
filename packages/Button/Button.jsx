import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { componentWithName, or } from '@tds/util-prop-types'
import { borders, forms } from '@tds/shared-styles'
import { medium, boldFont } from '@tds/shared-typography'
import { colorPrimary, colorSecondary, colorWhite, colorText } from '@tds/core-colours'
import { media } from '@tds/core-responsive'

import { safeRest } from '@tds/util-helpers'
import { warn } from '../../shared/utils/warn'

const preventDisabling = ({ disabled, ...props }) => {
  if (disabled) {
    warn('Button', 'Buttons are not able to be disabled.')
  }

  return props
}
const baseButton = {
  margin: 0,
  padding: '0 2rem',
  cursor: 'pointer',
  background: 'none',
  transition: 'background 0.2s',
  display: 'flex',
  width: '100%',
  alignItems: 'center',
  justifyContent: 'center',

  ...media.from('md').css({
    display: 'inline-flex',
    width: 'auto',
    minWidth: '180px',
  }),
}
export const StyledButton = styled.button(
  borders.none,
  borders.rounded,
  medium,
  boldFont,
  forms.buttonHeight,
  forms.font,
  baseButton,
  ({ variant }) => {
    let backgroundColor
    let color
    const hover = {
      boxShadow: '0 0 0 1px',
    }

    if (variant === 'primary') {
      backgroundColor = colorPrimary
      color = colorWhite
      hover.backgroundColor = colorWhite
      hover.color = colorPrimary
    } else if (variant === 'secondary') {
      backgroundColor = colorSecondary
      color = colorWhite
      hover.backgroundColor = colorWhite
      hover.color = colorSecondary
    } else {
      backgroundColor = colorWhite
      color = colorText
      hover.backgroundColor = 'transparent'
      hover.color = colorWhite
    }

    return {
      backgroundColor,
      color,
      '&:hover': hover,
    }
  },
  ({ fullwidth }) => {
    let width
    if (fullwidth) {
      return {
        width: '100%',
        ...media.from('md').css({ width: '100%' }),
      }
    }
    return {
      width,
    }
  }
)

export const ButtonTextWrapper = styled.span({
  width: '100%',
})

/**
 * @version ./package.json
 */
const Button = forwardRef(({ type, variant, fullwidth, children, ...rest }, ref) => {
  const restNoDisabled = preventDisabling(rest)

  return (
    <StyledButton
      {...safeRest(restNoDisabled)}
      variant={variant}
      fullwidth={fullwidth}
      type={type}
      ref={ref}
    >
      <ButtonTextWrapper>{children}</ButtonTextWrapper>
    </StyledButton>
  )
})

Button.displayName = 'Button'

Button.propTypes = {
  /**
   * The HTML button type.
   */
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  /**
   * The style.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted']),
  /**
   * The label. It can include the `A11yContent` component or strings.
   */
  children: or([PropTypes.string, componentWithName('A11yContent')]).isRequired,
  /**
   * If `true`, sets `Button` to 100% width.
   * @since 2.2.0
   */
  fullwidth: PropTypes.bool,
}
Button.defaultProps = {
  type: 'button',
  variant: 'primary',
  fullwidth: false,
}

export default Button
