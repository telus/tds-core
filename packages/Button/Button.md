### Minimal usage

Provide a function as the `onClick` prop to perform an action when clicked. **Avoid using a button if navigation
is the primary action, as a [Link](#link) is more appropriate.**

### Usage criteria

- Use buttons to move through a transaction
- Aim to use only one primary button per page
- Keep the text short and able to fit on a single line (our recommendation is 2-24 characters)
- Make sure the button text describes an action
- Buttons should not be disabled. Use inline (on blur) error messaging to provide feedback when the form is invalid

By default, Buttons will be displayed in the `primary` variant. Use primary buttons for the main action on a page or
in a form.

```jsx
<Button>Submit</Button>
```

Specify the `variant` to create a button for secondary actions.

```jsx
<Button variant="secondary">Find out more</Button>
```

### Sizes

All buttons are inline, with a minimum width of 180px for screens larger than 768px. They will occupy 100% width of their
parent's width at the small viewport and below. Resize your browser window to see this behaviour.

### Placing buttons on dark backgrounds

Use the `inverted` button on top of a dark background (TELUS approved colours or images).

Use this variant with caution. There will be accessibility issues if the colour contrast of the image and the button
text is too low in the hover state.

```jsx { "props": { "className": "docs_purple-block" } }
<Button variant="inverted">Get started</Button>
```

### Using A11yContent

Use the `A11yContent` component to create invisible text that is read out loud by screen readers.

```jsx
<Button>
  Shop <A11yContent>iPhone</A11yContent>
</Button>
```
