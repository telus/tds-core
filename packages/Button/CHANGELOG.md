# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [2.2.0](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.14...@tds/core-button@2.2.0) (2020-07-22)


### Features

* **core-button:** add span as a valid child prop-type ([56f3838](https://github.com/telusdigital/tds/commit/56f38383ee25dec0fea864d816df8305ea4f2b79))





## [2.1.14](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.13...@tds/core-button@2.1.14) (2020-02-03)

**Note:** Version bump only for package @tds/core-button





## [2.1.13](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.12...@tds/core-button@2.1.13) (2020-01-27)

**Note:** Version bump only for package @tds/core-button





## [2.1.12](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.11...@tds/core-button@2.1.12) (2020-01-23)

**Note:** Version bump only for package @tds/core-button





## [2.1.11](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.10...@tds/core-button@2.1.11) (2020-01-22)

**Note:** Version bump only for package @tds/core-button





## [2.1.10](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.9...@tds/core-button@2.1.10) (2019-10-09)

**Note:** Version bump only for package @tds/core-button





## [2.1.9](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.8...@tds/core-button@2.1.9) (2019-10-04)

**Note:** Version bump only for package @tds/core-button





## [2.1.8](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.7...@tds/core-button@2.1.8) (2019-09-12)

**Note:** Version bump only for package @tds/core-button





## [2.1.7](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.6...@tds/core-button@2.1.7) (2019-09-06)

**Note:** Version bump only for package @tds/core-button





## [2.1.6](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.5...@tds/core-button@2.1.6) (2019-08-22)

**Note:** Version bump only for package @tds/core-button





## [2.1.5](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.4...@tds/core-button@2.1.5) (2019-08-09)

**Note:** Version bump only for package @tds/core-button





## [2.1.4](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.3...@tds/core-button@2.1.4) (2019-08-01)

**Note:** Version bump only for package @tds/core-button





## [2.1.3](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.2...@tds/core-button@2.1.3) (2019-07-30)

**Note:** Version bump only for package @tds/core-button





## [2.1.2](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.1...@tds/core-button@2.1.2) (2019-07-25)

**Note:** Version bump only for package @tds/core-button





## [2.1.1](https://github.com/telusdigital/tds/compare/@tds/core-button@2.1.0...@tds/core-button@2.1.1) (2019-06-25)

**Note:** Version bump only for package @tds/core-button





# [2.1.0](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.8...@tds/core-button@2.1.0) (2019-06-13)


### Features

* **core-button:** add forwardRef ([fd8f181](https://github.com/telusdigital/tds/commit/fd8f181))





## [2.0.8](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.7...@tds/core-button@2.0.8) (2019-06-04)

**Note:** Version bump only for package @tds/core-button

## [2.0.7](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.6...@tds/core-button@2.0.7) (2019-05-14)

### Bug Fixes

- **core-button:** adding a span inside basic button to center text ([9c85a38](https://github.com/telusdigital/tds/commit/9c85a38))
- **core-button:** convert span into styled component ([3f7e377](https://github.com/telusdigital/tds/commit/3f7e377))
- **core-button:** moved styles inside span component ([0ffc41e](https://github.com/telusdigital/tds/commit/0ffc41e))

## [2.0.6](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.5...@tds/core-button@2.0.6) (2019-05-13)

**Note:** Version bump only for package @tds/core-button

## [2.0.5](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.4...@tds/core-button@2.0.5) (2019-05-07)

**Note:** Version bump only for package @tds/core-button

## [2.0.4](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.3...@tds/core-button@2.0.4) (2019-05-01)

**Note:** Version bump only for package @tds/core-button

## [2.0.3](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.2...@tds/core-button@2.0.3) (2019-05-01)

**Note:** Version bump only for package @tds/core-button

## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.1...@tds/core-button@2.0.2) (2019-04-04)

**Note:** Version bump only for package @tds/core-button

## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-button@2.0.0...@tds/core-button@2.0.1) (2019-03-28)

### Bug Fixes

- **core-button:** remove @tds/shared-base-button from deps ([7b4783f](https://github.com/telusdigital/tds/commit/7b4783f))

# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.10...@tds/core-button@2.0.0) (2019-03-28)

### Features

- **core-button:** upgrade to styled-components ([0ff9dba](https://github.com/telusdigital/tds/commit/0ff9dba))

### BREAKING CHANGES

- **core-button:** upgrade to styled-components

## [1.1.10](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.9...@tds/core-button@1.1.10) (2019-03-28)

**Note:** Version bump only for package @tds/core-button

## [1.1.9](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.8...@tds/core-button@1.1.9) (2019-03-20)

**Note:** Version bump only for package @tds/core-button

## [1.1.8](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.7...@tds/core-button@1.1.8) (2019-03-20)

**Note:** Version bump only for package @tds/core-button

## [1.1.7](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.6...@tds/core-button@1.1.7) (2019-03-18)

**Note:** Version bump only for package @tds/core-button

## [1.1.6](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.5...@tds/core-button@1.1.6) (2019-03-11)

**Note:** Version bump only for package @tds/core-button

## [1.1.5](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.4...@tds/core-button@1.1.5) (2019-03-05)

**Note:** Version bump only for package @tds/core-button

## [1.1.4](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.3...@tds/core-button@1.1.4) (2019-02-06)

**Note:** Version bump only for package @tds/core-button

## [1.1.3](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.2...@tds/core-button@1.1.3) (2019-01-10)

### Bug Fixes

- **core-button:** replace airbnb proptypes with util-prop-types ([8ec7352](https://github.com/telusdigital/tds/commit/8ec7352))

<a name="1.1.2"></a>

## [1.1.2](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.1...@tds/core-button@1.1.2) (2018-12-19)

**Note:** Version bump only for package @tds/core-button

<a name="1.1.1"></a>

## [1.1.1](https://github.com/telusdigital/tds/compare/@tds/core-button@1.1.0...@tds/core-button@1.1.1) (2018-07-23)

**Note:** Version bump only for package @tds/core-button

<a name="1.1.0"></a>

# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-button@1.0.1...@tds/core-button@1.1.0) (2018-07-17)

### Bug Fixes

- **shared-base-button:** add span wrapper to support spaces ([b9a1984](https://github.com/telusdigital/tds/commit/b9a1984))

### Features

- **core-button:** add A11yContent props ([d8d540d](https://github.com/telusdigital/tds/commit/d8d540d))
- **core-button:** support A11yContent as a child component ([aec9617](https://github.com/telusdigital/tds/commit/aec9617))
- **core-button-link:** add A11yContent props ([7867a41](https://github.com/telusdigital/tds/commit/7867a41))

<a name="1.0.1"></a>

## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-button@1.0.0...@tds/core-button@1.0.1) (2018-04-02)

### Bug Fixes

- **shared-base-button:** change sizing class name ([449e4b4](https://github.com/telusdigital/tds/commit/449e4b4))
- **shared-base-button:** refactor for sass based responsiveness ([a17f6ac](https://github.com/telusdigital/tds/commit/a17f6ac))
