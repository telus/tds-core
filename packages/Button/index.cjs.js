const Button = require('./dist/index.cjs')

module.exports = Button
