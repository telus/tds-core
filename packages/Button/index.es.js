export * from './dist/index.es'
export { default } from './dist/index.es'
