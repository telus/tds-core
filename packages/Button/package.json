{
  "name": "@tds/core-button",
  "version": "1.1.1",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telusdigital/tds.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telusdigital/tds/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": ">=15",
    "react-dom": ">=15"
  },
  "dependencies": {
    "prop-types": "^15.5.10"
  },
  "devDependencies": {
    "@tds/core-colours": "^1.0.0",
    "@tds/shared-base-button": "^1.2.1",
    "airbnb-prop-types": "2.11.0"
  }
}
