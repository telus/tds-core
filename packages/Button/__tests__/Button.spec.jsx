import React from 'react'
import { render, mount, shallow } from 'enzyme'

import A11yContent from '@tds/core-a11y-content'

import { warn } from '../../../shared/utils/warn'

import Button from '../Button'

jest.mock('../../../shared/utils/warn')

describe('Button', () => {
  const doMount = (overrides = {}) => {
    const button = mount(<Button {...overrides}>Submit</Button>)

    return button.find('button')
  }
  const doShallow = (overrides = {}) => {
    const button = shallow(<Button {...overrides}>Submit</Button>)

    return button
  }

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders', () => {
    const button = render(<Button>Submit</Button>)

    expect(button).toMatchSnapshot()
  })

  it('has one of the HTML button types', () => {
    let button = doShallow()
    expect(button).toHaveProp('type', 'button')

    button = doShallow({ type: 'reset' })
    expect(button).toHaveProp('type', 'reset')
  })

  it('can be presented as one of the allowed variants', () => {
    let button = doMount()
    expect(button).toHaveClassName('primary')

    button = doMount({ variant: 'secondary' })
    expect(button).toHaveClassName('secondary')
  })

  it('can not be disabled', () => {
    const button = doShallow({ disabled: true })

    expect(button).not.toHaveProp('disabled')
    expect(warn).toHaveBeenCalled()
  })

  it('passes additional attributes to button element', () => {
    const button = doShallow({ id: 'the-button', tabIndex: 1 })

    expect(button).toHaveProp('id', 'the-button')
    expect(button).toHaveProp('tabIndex', 1)
  })

  describe('A11yContent', () => {
    it('connects to Button', () => {
      const button = shallow(
        <Button>
          Go home
          <A11yContent>testing</A11yContent>
        </Button>
      )

      expect(button).toContainReact(<A11yContent>testing</A11yContent>)
    })
  })

  it('does not allow custom CSS', () => {
    const button = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(button).not.toHaveProp('className', 'my-custom-class')
    expect(button).not.toHaveProp('style')
  })
})
