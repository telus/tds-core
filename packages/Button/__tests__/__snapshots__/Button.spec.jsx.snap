// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button can be presented as one of the allowed variants 1`] = `
.c0 {
  border-width: 0;
  border-radius: 4px;
  word-wrap: break-word;
  font-size: 1rem;
  -webkit-letter-spacing: -0.8px;
  -moz-letter-spacing: -0.8px;
  -ms-letter-spacing: -0.8px;
  letter-spacing: -0.8px;
  line-height: 1.5;
  word-wrap: break-word;
  font-weight: 700;
  font-family: "TELUS-Web","Helvetica Neue",Helvetica,Arial,sans-serif;
  background-color: #4b286d;
  color: #fff;
}

.c0:hover {
  box-shadow: 0 0 0 1px;
  background-color: #fff;
  color: #4b286d;
}

.c1 {
  width: 100%;
}

<button
  class="c0"
  type="button"
>
  <span
    class="c1"
  >
    Submit
  </span>
</button>
`;

exports[`Button renders 1`] = `
.c0 {
  border-width: 0;
  border-radius: 4px;
  word-wrap: break-word;
  font-size: 1rem;
  -webkit-letter-spacing: -0.8px;
  -moz-letter-spacing: -0.8px;
  -ms-letter-spacing: -0.8px;
  letter-spacing: -0.8px;
  line-height: 1.5;
  word-wrap: break-word;
  font-weight: 700;
  font-family: "TELUS-Web","Helvetica Neue",Helvetica,Arial,sans-serif;
  background-color: #2B8000;
  color: #fff;
}

.c0:hover {
  box-shadow: 0 0 0 1px;
  background-color: #fff;
  color: #2B8000;
}

.c1 {
  width: 100%;
}

<button
  class="c0"
  type="button"
>
  <span
    class="c1"
  >
    Submit
  </span>
</button>
`;
